hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/MergeHiveSchemaWithAvro.java

```diff
@@ -0,0 +1,197 @@
+    }
+  }
+
+  /**
+   * Reorders an option schema so that the type of the provided default value is the first type in the option schema
+   *
+   * e.g. If the schema is (NULL, INT) and the default value is 1, the returned schema is (INT, NULL)
+   * If the schema is not an option schema or if there is no default value, schema is returned as-is
+   */
+  private Schema reorderOptionIfRequired(Schema schema, Object defaultValue) {
+    if (AvroSchemaUtil.isOptionSchema(schema) && defaultValue != null) {
+      boolean isNullFirstOption = schema.getTypes().get(0).getType() == Schema.Type.NULL;
+      if (isNullFirstOption && defaultValue.equals(JsonProperties.NULL_VALUE)) {
+        return schema;
+      } else {
+        return Schema.createUnion(schema.getTypes().get(1), schema.getTypes().get(0));
+      }
+    } else {
+      return schema;
+    }
+  }
+
+  @Override
+  public Schema list(ListTypeInfo list, Schema partner, Schema elementResult) {
+    // if there was no matching Avro list, or if matching Avro list was an option, return an optional list
+    boolean shouldResultBeOptional = partner == null || AvroSchemaUtil.isOptionSchema(partner);
+    Schema result = Schema.createArray(elementResult);
+    return shouldResultBeOptional ? AvroSchemaUtil.toOption(result) : result;
+  }
+
+  @Override
+  public Schema map(MapTypeInfo map, Schema partner, Schema keyResult, Schema valueResult) {
+    Preconditions.checkArgument(extractIfOption(keyResult).getType() == Schema.Type.STRING,
+        "Map keys should always be non-nullable strings. Found: %s", keyResult);
+    // if there was no matching Avro map, or if matching Avro map was an option, return an optional map
+    boolean shouldResultBeOptional = partner == null || AvroSchemaUtil.isOptionSchema(partner);
+    Schema result = Schema.createMap(valueResult);
+    return shouldResultBeOptional ? AvroSchemaUtil.toOption(result) : result;
+  }
+
+  @Override
+  public Schema primitive(PrimitiveTypeInfo primitive, Schema partner) {
+    boolean shouldResultBeOptional = partner == null || AvroSchemaUtil.isOptionSchema(partner);
+    Schema hivePrimitive = HiveTypeUtil.visit(primitive, hiveToAvro);
+    // if there was no matching Avro primitive, use the Hive primitive
+    Schema result = partner == null ? hivePrimitive : checkCompatibilityAndPromote(hivePrimitive, partner);
+    return shouldResultBeOptional ? AvroSchemaUtil.toOption(result) : result;
+  }
+
+  private Schema checkCompatibilityAndPromote(Schema schema, Schema partner) {
+    // TODO: Check if schema is compatible with partner
+    //       Also do type promotion if required, schema = string & partner = enum, schema = bytes & partner = fixed, etc
+    return schema;
+  }
+
+  /**
+   * A {@link PartnerAccessors} which matches the requested field from a partner Avro struct by case insensitive
+   * field name match
+   */
+  private static class AvroPartnerAccessor implements PartnerAccessors<Schema, Schema.Field> {
```

Comment:
Nit: `AvroPartnerAccessor implements PartnerAccessor` (without `s` at the end) reads better in my opinion. We can do that after check-in as well.

Reply:
Makes sense